nomenclature/codelist.py

```diff
@@ -339,28 +339,28 @@ def check_illegal_characters(self, config: NomenclatureConfig) -> dict[str, Code
         illegal = ["{", "}"] + config.illegal_characters
         errors = ErrorCollector()
 
-        def _check_string(value):
+        def _check_string(attr, value):
             if isinstance(value, str):
-                if any(char in value for char in illegal):
+                if found := set(illegal).intersection(value):
+                    found = "', '".join(sorted(found))
                     errors.append(
                         ValueError(
-                            f"Unexpected character in {self.name}: '{code.name}'."
+                            f"Unexpected character(s) '{found}' in {self.name}.{attr} of: '{code.name}'."
                             " Check for illegal characters and/or if tags were spelled correctly."
                         )
                     )
             elif isinstance(value, dict):
-                for k in value.keys():
-                    _check_string(k)
-                for v in value.values():
-                    _check_string(v)
+                for k, v in value.items():
+                    _check_string(k, k)
+                    _check_string(k, v)
             elif isinstance(value, list):
                 for item in value:
-                    _check_string(item)
+                    _check_string(attr, item)
 
         for code in self.mapping.values():
             if not code.repository:
```

Comment:
Not part of this PR, but I'd suggest renaming this attribute to `from_external_repository`. This would make the code easier to read:
```suggestion
            if not code.from_external_repository:
```